src/datastore/pg-write-store.ts

```diff
@@ -2716,27 +2891,102 @@ export class PgWriteStore extends PgStore {
+          ON CONFLICT (address, token)
+          DO UPDATE
+          SET balance = ft_balances.balance + EXCLUDED.balance
+          RETURNING ft_balances.address
+        )
+        SELECT 
+          (SELECT COUNT(*)::int FROM updated_events) AS updated_events_count,
+          (SELECT COUNT(*)::int FROM update_balances) AS update_balances_count
       `;
+      const updateCount = stxResults[0]?.updated_events_count ?? 0;
       if (canonical) {
-        updatedEntities.markedCanonical.stxEvents += stxResults.count;
+        updatedEntities.markedCanonical.stxEvents += updateCount;
       } else {
-        updatedEntities.markedNonCanonical.stxEvents += stxResults.count;
+        updatedEntities.markedNonCanonical.stxEvents += updateCount;
       }
     });
     q.enqueue(async () => {
-      const ftResult = await sql`
-        UPDATE ft_events
-        SET canonical = ${canonical}
-        WHERE index_block_hash = ${indexBlockHash} AND canonical != ${canonical}
+      const ftResult = await sql<{ updated_events_count: number; update_balances_count: number }[]>`
+        WITH updated_events AS (
+          UPDATE ft_events
+          SET canonical = ${canonical}
+          WHERE index_block_hash = ${indexBlockHash} AND canonical != ${canonical}
+          RETURNING sender, recipient, amount, asset_event_type_id, asset_identifier, canonical
+        ),
+        event_changes AS (
+          SELECT address, asset_identifier, SUM(balance_change) AS balance_change
+          FROM (
+              SELECT sender AS address, asset_identifier,
+                SUM(CASE WHEN canonical THEN -amount ELSE amount END) AS balance_change
+              FROM updated_events
+              WHERE asset_event_type_id IN (1, 3) -- Transfers and Burns affect the sender's balance
+              GROUP BY sender, asset_identifier
+            UNION ALL
+              SELECT recipient AS address, asset_identifier,
+                SUM(CASE WHEN canonical THEN amount ELSE -amount END) AS balance_change
+              FROM updated_events
+              WHERE asset_event_type_id IN (1, 2) -- Transfers and Mints affect the recipient's balance
+              GROUP BY recipient, asset_identifier
+          ) AS subquery
+          GROUP BY address, asset_identifier
+        ),
+        update_balances AS (
+          INSERT INTO ft_balances (address, token, balance)
+          SELECT ec.address, ec.asset_identifier, ec.balance_change
+          FROM event_changes ec
+          ON CONFLICT (address, token)
+          DO UPDATE
+          SET balance = ft_balances.balance + EXCLUDED.balance
+          RETURNING ft_balances.address
+        )
+        SELECT 
+          (SELECT COUNT(*)::int FROM updated_events) AS updated_events_count,
+          (SELECT COUNT(*)::int FROM update_balances) AS update_balances_count
       `;
+      const updateCount = ftResult[0]?.updated_events_count ?? 0;
+      if (updateCount > 0) {
+        console.log('wat');
```

Comment:
This intended to be kept?

Reply:
Also noticed a lot of `update_balances_count` being returned in the end, but never being used — wasn't sure what these are for?

Reply:
Good catch!